src/datastore/postgres-store.ts

```diff
@@ -4745,6 +4809,44 @@ export class PgDataStore
     });
   }
 
+  /**
+   * Refreshes the `nft_custody` and `nft_custody_unanchored` materialized views if necessary.
+   * @param client - DB client
+   * @param txs - Transaction event data
+   * @param unanchored - If this refresh is requested from a block or microblock
+   */
+  async refreshNftCustody(
+    client: ClientBase,
+    txs: DataStoreTxEventData[],
+    unanchored: boolean = false
+  ) {
+    const newNftEventCount = txs
+      .map(tx => tx.nftEvents.length)
+      .reduce((prev, cur) => prev + cur, 0);
+    if (newNftEventCount > 0) {
+      // Always refresh unanchored view since even if we're in a new anchored block we should update the
+      // unanchored state to the current one.
+      await this.refreshMaterializedView(client, 'nft_custody_unanchored');
+      if (!unanchored) {
+        await this.refreshMaterializedView(client, 'nft_custody');
+      }
+    } else if (!unanchored) {
+      // Even if we didn't receive new NFT events in a new anchor block, we should check if we need to
+      // update the anchored view to reflect any changes made by previous microblocks.
+      const outdated = await client.query<{ outdated: boolean }>(
+        `
+        WITH anchored_height AS (SELECT MAX(block_height) AS anchored FROM nft_custody),
+          unanchored_height AS (SELECT MAX(block_height) AS unanchored FROM nft_custody_unanchored)
+        SELECT unanchored > anchored AS outdated
+        FROM anchored_height CROSS JOIN unanchored_height
+        `
+      );
+      if (outdated) {
+        await this.refreshMaterializedView(client, 'nft_custody');
+      }
```

Comment:
The above code looks right. Are you confident the unanchored tests are good enough coverage or should we add some re-org tests as well?

Reply:
Good looking out. I added more tests and a fix for only considering NFT events from canonical txs.